### PR TITLE
[TR] Contributor guide

### PR DESCRIPTION
**Pull request by willy-ahva**
Contributor guide for Akeneo PIM project.

Inspired from Sylius guide : http://docs.sylius.org/en/latest/contributing/code/patches.html

### CODE REVIEW

**Comment by a2xchip** [comment]
@willy-ahva Rebasing and patching?

**Comment by a2xchip** [comment]
I was thinking about how to make project guide very fulfilled and stopped on idea that sylius guide is pretty good and enough.
The only idea i have now is too include more info about tests and testing practices and borrow django's structure [Contributing to Django](https://docs.djangoproject.com/en/1.9/internals/contributing/) mixing it with sylius information)))

**Comment by willy-ahva** [comment]
@a2xchip Django's documentation seems very "fat" and I think that a too big documentation could hind contributors to read it. I was more partisan to borrow the Sylius one that seems more concise and less repellent for a new comer. Don't hesitate to contribute on my PR or comment. :)

Can be discuss with @nidup

File: contributing/contribution_guide.rst [hunk 1]
```diff
@@ -0,0 +1,128 @@
+How to contribute ?
+===================
+
+As Akeneo PIM is an open-source project all contribution is very welcome.
```
Comment (a2xchip) [inline contributing/contribution_guide.rst]: As i think.
~ all contributions are very welcome 
or
~ any contribution is very welcome

File: contributing/contribution_guide.rst [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+http://docs.akeneo.com/latest/reference/best_practices/core/conventions.html and coding standard
+http://docs.akeneo.com/latest/reference/best_practices/core/standards.html to be sure to respect all our standard
+and avoid a lot of comments from core developers on your final Pull Request.
+
+For all contributions, please keep in mind that tests are as important as business code.
+Behavior of the application has to be tested with Behat.
+
+.. note::
+
+    How to setup Behat in Akeneo PIM http://docs.akeneo.com/latest/reference/best_practices/core/behat.html
```
Comment (a2xchip) [inline contributing/contribution_guide.rst]: May be we need to move Behat and Phpspec from **best practices** to **contribution** joining like Sylius and Django do?

File: contributing/contribution_guide.rst [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+Create a new branch
+
+.. code-block:: php
+
+    $ git checkout -b BRANCH_NAME master
+
+.. note::
+
+    How to name your branch ?
```
Comment (a2xchip) [inline contributing/contribution_guide.rst]: May be we can reffer to [git-flow](http://docs.akeneo.com/latest/developer_guide/installation/installation_workstation.html) practices in branching i.e
```
feature/ticket-num+ticket-name
bug/ticket-num+ticket-name
```
or implement something similar. As i see todays branch naming is not standardized yet. @nidup may help and express his thoughts about it.

File: contributing/contribution_guide.rst [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+.. code-block:: php
+
+    $ git commit
+
+Submitting a Pull Request
+-------------------------
+
+Simple rules to follow before submitting a PR :
+ - Squash irrelevant commits
```
Comment (a2xchip) [inline contributing/contribution_guide.rst]: [Github supports squash's](https://github.com/blog/2141-squash-your-commits). May be it is not so actual now.

**Comment by a2xchip** [comment]
@willy-ahva Right it is. I am talking not about information and its volume but the structure itself. Sylius real structure is so: 

**Code**
* Bugs
* Patches
* Security
* Behavior Driven Development
* Coding Standards
* Code Conventions
* Git
* License

**Documentation**
* Overview
* Format
* Documentation Standards

**Community**
* Support
* Statistics

File: contributing/contribution_guide.rst (hunk 2 above)
Comment (willy-ahva) [inline contributing/contribution_guide.rst]: Totally agree for tests concern. It should be put in contribution part.

File: contributing/contribution_guide.rst (hunk 3 above)
Comment (nidup) [inline contributing/contribution_guide.rst]: Core team use internal ticket number (Jira), they looks like PIM-1234.
This convention seems ok for contribution which references github issues.

File: contributing/contribution_guide.rst (hunk 2 above)
Comment (nidup) [inline contributing/contribution_guide.rst]: Best practices and contribution section looks pretty duplicated in fact :) 
Ok to move these pages to contributing section and add a link from the following page to the contributing chapter http://docs.akeneo.com/latest/reference/best_practices/core.html

**Comment by nidup** [comment]
I really like the Sylius guide which is quite light and well organized, we can take inspiration from it for a first version then improve it and iterate on this topic.

**Comment by nidup** [comment]
Hi @pjedrzejewski !

Is it a problem for you if we take a large inspiration of the contributing section of Sylius documentation (http://docs.sylius.org/en/latest/contributing/index.html) to integrate it in the Akeneo PIM documentation?
Your doc is really great and the concerns / rules are almost the same to contribute on Sylius or Akeneo.
Obviously, we'll add a special thank you notice in our doc with a link to the Sylius doc :)
Don't hesitate to say us if it's a problem to take inspiration from this resource.
Thank you.

File: contributing/contribution_guide.rst (hunk 3 above)
Comment (a2xchip) [inline contributing/contribution_guide.rst]: @nidup The thing is only important is that all PR branches must be named using same rule. GH-1234 for github issues?

**Comment by a2xchip** [comment]
@nidup The Sylius licence allow it without any permissions but anyways Pavel's blessing is very welcome! ))))

File: contributing/contribution_guide.rst [hunk 5]
```diff
@@ -0,0 +1,393 @@
+Submitting a Patch
+==================
+
+As Akeneo PIM is an open-source project all contributions are very welcome.
```
Comment (rybus) [inline contributing/contribution_guide.rst]: project, (add a comma)

File: contributing/contribution_guide.rst [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+Patches are the best way to provide a bug fix or to propose enhancements.
+All features contributions are submitting to our Product Owner before to merge.
```
Comment (rybus) [inline contributing/contribution_guide.rst]: are submitted before being merged

File: contributing/contribution_guide.rst [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+Only them have to power to accept your contribution !
```
Comment (rybus) [inline contributing/contribution_guide.rst]: Only them have the power to accept your contribution! (no space before "!")

File: contributing/contribution_guide.rst [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+``s``. When you save, Git will start rebasing, and if successful, will ask
+you to edit the commit message, which by default is a listing of the commit
+messages of all the commits. When you are finished, execute the push command.
+
+.. _`Akeneo PIM repository`:            https://github.com/akeneo/pim-community-dev
+.. _ProGit:                             http://git-scm.com/book
+.. _GitHub:                             https://github.com/signup/free
+.. _`GitHub's Documentation`:           https://help.github.com/articles/ignoring-files
+.. _`documentation repository`:         https://github.com/akeneo/pim-docs
+.. _`Sylius documentation`:             http://docs.sylius.org/en/latest/contributing/index.html
```
Comment (juliensnz) [inline contributing/contribution_guide.rst]: wut ?

File: contributing/contribution_guide.rst [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+software:
+
+* Git
+* PHP version 5.4.4 ≤ version ≤ 5.6
+* MySQL 5.1 ≤ version ≤ 5.6
+
+Configure Git
+~~~~~~~~~~~~~
+
+Set up your user information with your real name and a working email address:
```
Comment (rybus) [inline contributing/contribution_guide.rst]: Setup

File: contributing/contribution_guide.rst [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+
+Note that documentation has been largely inspired from the `Sylius documentation`_. Thank them for the great work.
+
+Step 1: Setup your Environment
+------------------------------
+
+Install the Software Stack
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before working on Akeneo PIM, setup a Symfony2 friendly environment with the following
```
Comment (rybus) [inline contributing/contribution_guide.rst]: you have to set a Symfony2 friendly environment up (to setup is not a verb)

File: contributing/contribution_guide.rst (hunk 9 above)
Comment (rybus) [inline contributing/contribution_guide.rst]: Same notice about "to set up" verb.

File: contributing/contribution_guide.rst (hunk 8 above)
Comment (willy-ahva) [inline contributing/contribution_guide.rst]: I added a link to Sylius doc, ours is very inspiring from theirs.

File: contributing/contribution_guide.rst [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+    $ git config --global user.email you@example.com
+
+.. tip::
+
+    If you are new to Git, you are highly recommended to read the excellent and
+    free `ProGit`_ book.
+
+.. tip::
+
+    If your IDE creates configuration files inside the project's directory,
```
Comment (rybus) [inline contributing/contribution_guide.rst]: inside the directory of the project (no `'s` with non-human objects).

File: contributing/contribution_guide.rst [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+    you can use global ``.gitignore`` file (for all projects) or
+    ``.git/info/exclude`` file (per project) to ignore them. See
+    `Github's documentation`_.
+
+.. tip::
+
+    Windows users: when installing Git, the installer will ask what to do with
+    line endings, and suggests replacing all LF with CRLF. This is the wrong
```
Comment (rybus) [inline contributing/contribution_guide.rst]: and will suggest

File: contributing/contribution_guide.rst [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+      $ cd pim-community-dev
+      $ git remote add upstream git://github.com/akeneo/pim-community-dev.git
+
+Step 2: Work on your Patch
+--------------------------
+
+The License
+~~~~~~~~~~~
+
+Before you start, you must know that all the patches you are going to submit
```
Comment (rybus) [inline contributing/contribution_guide.rst]: all patches

File: contributing/contribution_guide.rst [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+.. code-block:: bash
+
+    $ git checkout -b BRANCH_NAME master
+
+.. tip::
+
+    Use a descriptive name for your branch (``issue_XXX`` where ``XXX`` is the
+    GitHub issue number is a good convention for bug fixes).
+
+The above checkout commands automatically switch the code to the newly created
```
Comment (rybus) [inline contributing/contribution_guide.rst]: command (no s) swtiches

File: contributing/contribution_guide.rst [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+branch (check the branch you are working on with ``git branch``).
```
Comment (rybus) [inline contributing/contribution_guide.rst]: you can check

File: contributing/contribution_guide.rst [hunk 16]
```diff
@@ -0,0 +1,393 @@
...
+
+Work on your Patch
+~~~~~~~~~~~~~~~~~~
+
+Before working on a contribution for an Akeneo repository, please read the following `code conventions`_
+and `coding standard`_ to be sure to respect all our standards.
```
Comment (rybus) [inline contributing/contribution_guide.rst]: to make sure

File: contributing/contribution_guide.rst [hunk 17]
```diff
@@ -0,0 +1,393 @@
...
+* For a bug fix contribution, please avoid any BC breaks. If a BC break can't be avoided add a comment and detail why.
+
+* For all contributions, tests are as important as business code.
+
+    - Behavior of the application has to be tested with Behat.
+    - Behavior of the business code has to be tested with PHPSpec.
+
+.. note::
+
+    We did a guide to `setup behat`_ in Akeneo PIM and you can check the `behat quick intro`_ on their documentation.
```
Comment (rybus) [inline contributing/contribution_guide.rst]: we made/we wrote

**Comment by rybus** [comment]
GTM after fixes

**Comment by pierallard** [comment]
GTM for the content (not english fluent heh)

File: contributing/contribution_guide.rst [hunk 18]
```diff
@@ -0,0 +1,392 @@
+Submitting a Patch
+==================
+
+As Akeneo PIM is an open-source project, all contributions are very welcome.
+Patches are the best way to provide a bug fix or to propose enhancements.
+All features contributions are submitted before being merge to our Product Owner.
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: being merged

**Comment by jjanvier** [comment]
Is it possible to apply all the fixes to the Sylius doc? it would be a way to say "thank you"

File: contributing/contribution_guide.rst [hunk 19]
```diff
@@ -0,0 +1,392 @@
...
+Only them have to power to accept your contribution!
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: the power

File: contributing/contribution_guide.rst [hunk 20]
```diff
@@ -0,0 +1,392 @@
...
+
+Note that documentation has been largely inspired from the `Sylius documentation`_. Thank them for the great work.
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: Thanks to them

File: contributing/contribution_guide.rst [hunk 21]
```diff
@@ -0,0 +1,392 @@
...
+
+Configure Git
+~~~~~~~~~~~~~
+
+Set your user information up with your real name and a working email address:
+
+.. code-block:: bash
+
+    $ git config --global user.name "Your Name"
+    $ git config --global user.email you@example.com
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: you@example.com with " " ?

File: contributing/contribution_guide.rst [hunk 22]
```diff
@@ -0,0 +1,392 @@
+Submitting a Patch
+==================
+
+As Akeneo PIM is an open-source project, all contributions are very welcome.
+Patches are the best way to provide a bug fix or to propose enhancements.
+All features contributions are submitted before being merge to our Product Owner.
+Only them have the power to accept your contribution!
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: they

File: contributing/contribution_guide.rst [hunk 23]
```diff
@@ -0,0 +1,392 @@
...
+The above checkout command automatically switches the code to the newly created
+branch (you can check the branch you are working on with ``git branch``).
+
+Work on your Patch
+~~~~~~~~~~~~~~~~~~
+
+Before working on a contribution for an Akeneo repository, please read the following `code conventions`_
+and `coding standard`_ to make sure to respect all our standards.
+
+When you work on a patch, please keep in mind :
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: mind:

File: contributing/contribution_guide.rst [hunk 24]
```diff
@@ -0,0 +1,392 @@
...
+    $ git commit
+
+* Create atomic and logical commits with a relevant message.
+
+* Squash irrelevant commits that are just about fixing coding standards or fixing typos in your own code.
+
+* Never fix coding standards in some existing code as it makes the code review
+  more difficult (submit CS fixes as a separate patch).
+
+It will help us to :
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: to:

File: contributing/contribution_guide.rst [hunk 25]
```diff
@@ -0,0 +1,392 @@
...
+      - Typically a hyphen or asterisk is used for the bullet,
+        preceded by a single space, with blank lines in
+        between, but conventions vary here
+
+Prepare your Patch for Submission
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When your patch is about a bug fix and we give you a reference to a ticket ``PIM-xxxx``.
+You have to add to the ``CHANGELOG-1-x.md`` file (1-x is the version of the PIM you want to contribute)
+under the BUG FIXES step of the next version the reference to the ticket and a description of the bug fixe.
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: bud fix.

File: contributing/contribution_guide.rst [hunk 26]
```diff
@@ -0,0 +1,392 @@
...
+
+Then, if you introduced BC Breaks in namespaces for example (but you should/must not), under the BC BREAK step add a
+description of the BC Break.
+Moreover, you have to add in ``UPGRADE-1.x.md`` a way to fix this BC Break in files (like sed command for example).
+To finish, if you introduced database BC Break, you have to add migration files in `upgrades/schema/`.
+In most of the cases use `php app/console doctrine:migrations:diff` is enough to create a database migration class
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: using

File: contributing/contribution_guide.rst [hunk 27]
```diff
@@ -0,0 +1,392 @@
...
+(see `Doctrine migration documentation`_) but sometimes you will have to do it by hand.
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: manually

File: contributing/contribution_guide.rst [hunk 28]
```diff
@@ -0,0 +1,392 @@
...
+.. code-block:: bash
+
+    $ git checkout -b BRANCH_NAME master
+
+.. tip::
+
+    Use a descriptive name for your branch (``issue_XXX`` where ``XXX`` is the
+    GitHub issue number is a good convention for bug fixes).
+
+The above checkout command automatically switches the code to the newly created
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: The checkout command above

Comment (LaureBro) [inline contributing/contribution_guide.rst]: To the new one created (not sure of the sentence)

File: contributing/contribution_guide.rst [hunk 29]
```diff
@@ -0,0 +1,392 @@
...
+    | Added Specs                       | [yes|no]
+    | Added Behats                      | [yes|no]
+    | Changelog updated                 | [yes|no]
+    | Review and 2 GTM                  | [yes|no]
+    | Migration script                  | [yes|no]
+    | Tech Doc                          | [yes|no]
+
+Some explanation for this Definition of Done :
+
+* "Added Specs" means phpspec have been written, every classes has its own PHPSpec or the existing one has been updated except controllers, form types, commands, doctrine entity (POPO), symfony semantic config.
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: class

File: contributing/contribution_guide.rst [hunk 30]
```diff
@@ -0,0 +1,392 @@
...
+
+* "Added Behats" means scenario have been written, for nominal and limit cases, internal api can also be tested through behat via commands (like query or updater).
+
+* "Changelog updated" means the bug fix line has been added (in case of bug) via an explicit sentence, all the BC break (with the last minor version) have been listed and, in case of improvement (functional or technical), a short description (prefixed by the issue number).
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: BC breaks

File: contributing/contribution_guide.rst [hunk 31]
```diff
@@ -0,0 +1,392 @@
...
+
+* "Review and 2 GTM" means the technical review has been done, comments have been fixed and at least two teammates have given a Good To Merge (GTM). Update it just before to merge.
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: before you merge

File: contributing/contribution_guide.rst [hunk 32]
```diff
@@ -0,0 +1,392 @@
...
+
+* "Migration script" means you changed the data model and you provided migration script allowing to migrate data from previous minor version to the coming one.
```
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: scripts... upcoming one.

File: contributing/contribution_guide.rst (hunk 25 above)
Comment (rybus) [inline contributing/contribution_guide.rst]: bug

Comment (rybus) [inline contributing/contribution_guide.rst]: :p

File: contributing/contribution_guide.rst (hunk 28 above)
Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: No, "the newly created branch" is good :)

File: contributing/contribution_guide.rst (hunk 8 above)
Comment (a2xchip) [inline contributing/contribution_guide.rst]: May be we can delete "Largely inspired note" from the beginning if we have one here

Comment (willy-ahva) [inline contributing/contribution_guide.rst]: This line is an alias for the link. I use it in the first paragraph. There is only one link ;)

File: contributing/contribution_guide.rst [hunk 33]
```diff
@@ -0,0 +1,392 @@
...
+    GitHub issue number is a good convention for bug fixes).
+
+The above checkout command automatically switches the code to the newly created
+branch (you can check the branch you are working on with ``git branch``).
+
+Work on your Patch
+~~~~~~~~~~~~~~~~~~
+
+Before working on a contribution for an Akeneo repository, please read the following `code conventions`_
+and `coding standard`_ to make sure to respect all our standards.
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: to make sure you respect

File: contributing/contribution_guide.rst [hunk 34]
```diff
@@ -0,0 +1,392 @@
...
+under the BUG FIXES step of the next version the reference to the ticket and a description of the bug fixe.
+
+Then, if you introduced BC Breaks in namespaces for example (but you should/must not), under the BC BREAK step add a
+description of the BC Break.
+Moreover, you have to add in ``UPGRADE-1.x.md`` a way to fix this BC Break in files (like sed command for example).
+To finish, if you introduced database BC Break, you have to add migration files in `upgrades/schema/`.
+In most of the cases use `php app/console doctrine:migrations:diff` is enough to create a database migration class
+(see `Doctrine migration documentation`_) but sometimes you will have to do it by hand.
+
+When your patch is not about a bug fix (when you add a new feature or change
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: **If** your patch... (**if **you add a new..)

File: contributing/contribution_guide.rst [hunk 35]
```diff
@@ -0,0 +1,392 @@
...
+    | Q                                 | A
+    | --------------------------------- | ---
+    | Added Specs                       | no
+    | Added Behats                      | no
+    | Changelog updated                 | yes
+    | Review and 2 GTM                  | no
+    | Migration script                  | no
+    | Tech Doc                          | no
+
+If you just submitted your PR for a bug fix with some BC Breaks in database, an example could now look as follows:
```
Comment (LaureBro) [inline contributing/contribution_guide.rst]: as follow:

Comment (MarieMinasyan) [inline contributing/contribution_guide.rst]: "as follows" is the correct version ;)

**Comment by MarieMinasyan** [comment]
GTM :) nice job!

**Comment by a2xchip** [comment]
As experienced Symfony developers we were looking on any project contribution section except Symfony once! Anyway sylius page is its beauty copy :smile: :smile: :smile: https://symfony.com/doc/current/contributing/code/patches.html#step-1-setup-your-environment

**Comment by willy-ahva** [comment]
Yeah, I did a PR to fix some typo on Sylius doc and Pawel told me about it :smile:

**Comment by pjedrzejewski** [comment]
Hey guys! We have taken the base guide from Symfony, after asking Ryan of course. :) We just mention that it is inspired by Symfony, feel free to do the same, but I think credit should go to their awesome team! :) We just adjusted it a bit for Sylius and added BDD stuff. ;)

**Comment by a2xchip** [comment]
@pjedrzejewski Great! I see it now! :-)

image

**Comment by a2xchip** [comment]
@willy-ahva We really need to do the same, and also mention great Sylius on that Behat related part.

**Comment by willy-ahva** [comment]
@a2xchip @pjedrzejewski I'll do a PR to thanks Symfony2 in our doc :)

**Pull request merged**